staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go

```diff
@@ -2269,3 +2261,299 @@ func TestLoadObjects(t *testing.T) {
 		})
 	}
 }
+
+func TestApplyObjects(t *testing.T) {
+	testdirs := []string{"testdata/prune/simple"}
+	for _, testdir := range testdirs {
+		t.Run(testdir, func(t *testing.T) {
+			tf := cmdtesting.NewTestFactory().WithNamespace("test")
+			defer tf.Cleanup()
+
+			codec := scheme.Codecs.LegacyCodec(scheme.Scheme.PrioritizedVersionsAllGroups()...)
+
+			scheme := runtime.NewScheme()
+
+			listMapping := map[schema.GroupVersionResource]string{
+				{Group: "", Version: "v1", Resource: "namespaces"}: "NamespaceList",
+			}
+
+			fakeDynamicClient := dynamicfakeclient.NewSimpleDynamicClientWithCustomListKinds(scheme, listMapping)
+			tf.FakeDynamicClient = fakeDynamicClient
+
+			tf.UnstructuredClient = &fake.RESTClient{
+				NegotiatedSerializer: resource.UnstructuredPlusDefaultContentConfig().NegotiatedSerializer,
+				Client: fake.CreateHTTPClient(func(req *http.Request) (*http.Response, error) {
+					method := req.Method
+
+					tokens := strings.Split(strings.TrimPrefix(req.URL.Path, "/"), "/")
+
+					if len(tokens) == 2 && tokens[0] == "namespaces" && method == "GET" {
+						name := tokens[1]
+						gvr := schema.GroupVersionResource{Version: "v1", Resource: "namespaces"}
+						ns, err := fakeDynamicClient.Tracker().Get(gvr, "", name)
+						if err != nil {
+							if apierrors.IsNotFound(err) {
+								return &http.Response{StatusCode: http.StatusNotFound, Header: cmdtesting.DefaultHeader()}, nil
+							}
+							t.Fatalf("error getting object: %v", err)
+						}
+						return &http.Response{StatusCode: http.StatusOK, Header: cmdtesting.DefaultHeader(), Body: cmdtesting.ObjBody(codec, ns)}, nil
+					}
+
+					if len(tokens) == 1 && tokens[0] == "namespaces" && method == "POST" {
```

Comment:
One thing I'm noticing in these tests now that I've written some myself is that they're simulating CSA. I think it's good to have one basic test that proves we work with CSA, and that would be deleted if/when CSA is someday removed. But I'd suggest that we write the rest of our tests against SSA, both because it is the future and because it is much simpler to simulate with fake clients!

Reply:
FYI the test clients don't support SSA yet either (IIUC), but I think antoine & co are working on it.

Reply:
Ah - I totally understand what's going on now.. I lost track of the thread here when we talked about it, I thought my comment was on one of your PRs.  Sorry for my confusion!

Totally agree we should be using SSA where we can :-)